File: src/core/HexMetrics.ts
import * as THREE from 'three';
import { TerrainType } from '../types';

// Core hex geometry constants
export const HexMetrics = {
  // Hex geometry
  outerRadius: 1.0,                           // Corner to center distance
  innerRadius: 1.0 * 0.866025404,             // Edge to center (outer * sqrt(3)/2)

  // Elevation
  elevationStep: 0.4,                         // Height per elevation level
  maxElevation: 8,
  minElevation: -2,
  waterLevel: 0,                              // Sea level

  // Terraces (Catlike Coding style)
  terracesPerSlope: 2,                        // Number of flat terraces per slope
  get terraceSteps(): number {                // Total interpolation steps (2*terracesPerSlope + 1)
    return this.terracesPerSlope * 2 + 1;
  },
  get horizontalTerraceStepSize(): number {   // Horizontal interpolation step size
    return 1.0 / this.terraceSteps;
  },
  get verticalTerraceStepSize(): number {     // Vertical interpolation step size
    return 1.0 / (this.terracesPerSlope + 1);
  },

  // Blend regions
  solidFactor: 0.8,                           // Inner solid hex portion
  blendFactor: 0.2,                           // Outer blend portion

  // Get the 6 corner positions for a hex
  getCorners(): THREE.Vector3[] {
    const corners: THREE.Vector3[] = [];
    for (let i = 0; i < 6; i++) {
      const angle = (Math.PI / 3) * i + Math.PI / 6; // Start at 30°
      corners.push(new THREE.Vector3(
        Math.cos(angle) * this.outerRadius,
        0,
        Math.sin(angle) * this.outerRadius
      ));
    }
    return corners;
  },

  // Get corner at specific index (with wrapping)
  getCorner(index: number): THREE.Vector3 {
    const corners = this.getCorners();
    return corners[((index % 6) + 6) % 6].clone();
  },
};

/**
 * Terrace interpolation - the key to Catlike Coding style terraces.
 * Horizontal interpolation is linear, vertical only changes on odd steps.
 * This creates flat "platforms" separated by short slopes.
 *
 * @param a Start position
 * @param b End position
 * @param step Current step (1 to terraceSteps)
 * @returns Interpolated position
 */
export function terraceLerp(a: THREE.Vector3, b: THREE.Vector3, step: number): THREE.Vector3 {
  // Horizontal interpolation is linear
  const h = step * HexMetrics.horizontalTerraceStepSize;

  // Vertical interpolation only changes on odd steps
  // step: 1 -> v = 1/3, step: 2 -> v = 1/3, step: 3 -> v = 2/3, step: 4 -> v = 2/3
  const v = Math.floor((step + 1) / 2) * HexMetrics.verticalTerraceStepSize;

  return new THREE.Vector3(
    a.x + (b.x - a.x) * h,
    a.y + (b.y - a.y) * v,
    a.z + (b.z - a.z) * h
  );
}

/**
 * Interpolate color along terrace (same as terraceLerp but for colors).
 */
export function terraceColorLerp(a: THREE.Color, b: THREE.Color, step: number): THREE.Color {
  const h = step * HexMetrics.horizontalTerraceStepSize;
  return new THREE.Color(
    a.r + (b.r - a.r) * h,
    a.g + (b.g - a.g) * h,
    a.b + (b.b - a.b) * h
  );
}

// Terrain colors - stylized low-poly palette
export const TerrainColors: Record<TerrainType, number> = {
  [TerrainType.Ocean]: 0x1a4c6e,      // Deep blue
  [TerrainType.Coast]: 0x2d8bc9,      // Light blue
  [TerrainType.Plains]: 0x7cb342,     // Grass green
  [TerrainType.Forest]: 0x2e7d32,     // Dark green
  [TerrainType.Hills]: 0x8d6e63,      // Brown
  [TerrainType.Mountains]: 0x757575,  // Gray
  [TerrainType.Snow]: 0xeceff1,       // White
  [TerrainType.Desert]: 0xe6c86e,     // Sand yellow
  [TerrainType.Tundra]: 0x90a4ae,     // Blue-gray
  [TerrainType.Jungle]: 0x1b5e20,     // Deep green
  [TerrainType.Savanna]: 0xc5a855,    // Golden brown
  [TerrainType.Taiga]: 0x4a635d,      // Dark teal-green
};

// Get THREE.Color from terrain type
export function getTerrainColor(terrain: TerrainType): THREE.Color {
  return new THREE.Color(TerrainColors[terrain]);
}

// Slightly vary a color for visual interest
export function varyColor(color: THREE.Color, amount: number = 0.05): THREE.Color {
  const hsl = { h: 0, s: 0, l: 0 };
  color.getHSL(hsl);
  hsl.l += (Math.random() - 0.5) * amount;
  hsl.l = Math.max(0, Math.min(1, hsl.l));
  return new THREE.Color().setHSL(hsl.h, hsl.s, hsl.l);
}

// Terrain type to numeric index for shader
const TerrainTypeIndices: Record<TerrainType, number> = {
  [TerrainType.Ocean]: 0,
  [TerrainType.Coast]: 1,
  [TerrainType.Plains]: 2,
  [TerrainType.Forest]: 3,
  [TerrainType.Hills]: 4,
  [TerrainType.Mountains]: 5,
  [TerrainType.Snow]: 6,
  [TerrainType.Desert]: 7,
  [TerrainType.Tundra]: 8,
  [TerrainType.Jungle]: 9,
  [TerrainType.Savanna]: 10,
  [TerrainType.Taiga]: 11,
};

// Get numeric index for terrain type (for shaders)
export function getTerrainTypeIndex(terrain: TerrainType): number {
  return TerrainTypeIndices[terrain];
}

File: results/csharp_converted/hex_metrics.cs
using Godot;
using Godot.Collections;


//# Core hex geometry constants and utilities

//# Matches web/src/core/HexMetrics.ts
// Hex geometry
[GlobalClass]
public partial class HexMetrics : Godot.RefCounted
{
	public const double OUTER_RADIUS = 1.0;
	// Corner to center distance
	public const double INNER_RADIUS = OUTER_RADIUS * 0.866025404;

	// Edge to center (outer * sqrt(3)/2)
	// Elevation - sea level system (no negative elevations)
	public const double ELEVATION_STEP = 0.4;
	// Height per elevation level
	public const int MIN_ELEVATION = 0;
	// Ocean floor (deepest water)
	public const int SEA_LEVEL = 4;
	// Water surface elevation (water is 0-4)
	public const int LAND_MIN_ELEVATION = 5;
	// Minimum land elevation (always 1 above sea level)
	public const int MAX_ELEVATION = 13;

	// Highest land (8 levels above land min)
	// Terraces (Catlike Coding style)
	public const int TERRACES_PER_SLOPE = 2;


	// Number of flat terraces per slope
	public static int GetTerraceSteps()
	{
		return TERRACES_PER_SLOPE * 2 + 1;
	}


	public static double GetHorizontalTerraceStepSize()
	{
		return 1.0 / GetTerraceSteps();
	}


	public static double GetVerticalTerraceStepSize()
	{
		return 1.0 / (TERRACES_PER_SLOPE + 1);
	}


	// Blend regions (Catlike Coding style)
	public const double SOLID_FACTOR = 0.8;
	// Inner solid hex portion
	public const double BLEND_FACTOR = 0.2;


	// Outer blend portion (edge/corner connections)
	//# Get the 6 corner positions for a hex (flat-topped, starting at 30 degrees)
	public static Array<Vector3> GetCorners()
	{
		var corners = new Array{};
		foreach(int i in GD.Range(6))
		{
			var angle = (Mathf.Pi / 3.0) * i + Mathf.Pi / 6.0;
			// Start at 30 degrees


			corners.Append(new Vector3(, Mathf.Cos(angle) * OUTER_RADIUS, 0, Mathf.Sin(angle) * OUTER_RADIUS));
		}
		return corners;
	}


	//# Get corner at specific index (with wrapping)
	public static Vector3 GetCorner(int index)
	{
		var corners = GetCorners();
		return corners[((index % 6) + 6) % 6];


		//# Terrace interpolation - the key to Catlike Coding style terraces

	}//# Horizontal interpolation is linear, vertical only changes on odd steps
	public static Vector3 TerraceLerp(Vector3 a, Vector3 b, int step)
	{
		var h = step * GetHorizontalTerraceStepSize();
		var v = Mathf.Floor((step + 1) / 2.0) * GetVerticalTerraceStepSize();


		return new Vector3(, a.X + (b.X - a.X) * h, a.Y + (b.Y - a.Y) * v, a.Z + (b.Z - a.Z) * h);
	}


	//# Interpolate color along terrace
	public static Color TerraceColorLerp(Color a, Color b, int step)
	{
		var h = step * GetHorizontalTerraceStepSize();
		return a.Lerp(b, h);
	}


}